File: packages/webhooks/test/gate.test.ts
import { describe, expect, test } from 'bun:test';
import type { Operation, Store } from '@anyonce/core';
import { MemoryStore } from '@anyonce/core';
import type { Problem } from '@anyonce/core/http';
import { markVerified, webhookReceiver } from '../src/index';

function countingStore(): { store: Store; begins: Operation[] } {
  const inner = new MemoryStore();
  const begins: Operation[] = [];
  const store: Store = {
    begin(op, opts) {
      begins.push(op);
      return inner.begin(op, opts);
    },
    complete: (op, fence, result, now) => inner.complete(op, fence, result, now),
    abandon: (op, fence) => inner.abandon(op, fence),
    get: (op, now) => inner.get(op, now),
    purge: (now) => inner.purge(now),
  };
  return { store, begins };
}

function delivery(
  body = '{"a":1}',
  headers: Record<string, string> = { 'webhook-id': 'msg_1' },
): Request {
  return new Request('https://example.test/hooks/stripe', { method: 'POST', headers, body });
}

describe('verification gate', () => {
  test('REQ-WH-2: a receiver with neither verify nor verifiedMarker answers 500 configuration-error and never calls begin', async () => {
    const { store, begins } = countingStore();
    const messages: string[] = [];
    const handler = webhookReceiver({ store, logger: (m) => messages.push(m) })(
      async () => new Response('handled'),
    );
    const res = await handler(delivery());
    expect(res.status).toBe(500);
    expect(res.headers.get('Content-Type')).toBe('application/problem+json');
    const body = (await res.json()) as {
      code: string;
      status: number;
      title: string;
      detail: string;
    };
    expect(body.code).toBe('configuration-error');
    expect(body.status).toBe(500);
    expect(body.title).toBe(
      'The webhook endpoint could not establish that this delivery is genuine',
    );
    expect(body.detail).toBe('no verify callback or verifiedMarker is configured');
    expect(begins).toEqual([]);
    expect(messages).toHaveLength(1);
  });

  test('REQ-WH-2: the configuration error logs once however many requests arrive', async () => {
    const { store } = countingStore();
    const messages: string[] = [];
    const handler = webhookReceiver({ store, logger: (m) => messages.push(m) })(
      async () => new Response('handled'),
    );
    await handler(delivery());
    await handler(delivery());
    await handler(delivery());
    expect(messages).toHaveLength(1);
    expect(messages[0]).not.toContain('msg_1');
  });

  test('REQ-WH-2: a verify callback that returns false answers 401 signature-invalid and never calls begin', async () => {
    const { store, begins } = countingStore();
    const handler = webhookReceiver({ store, verify: () => false })(
      async () => new Response('handled'),
    );
    const res = await handler(delivery());
    expect(res.status).toBe(401);
    expect(((await res.json()) as { code: string }).code).toBe('signature-invalid');
    expect(begins).toEqual([]);
  });

  test('REQ-WH-2: a verify callback that throws is 500 configuration-error and never calls begin', async () => {
    const { store, begins } = countingStore();
    const messages: string[] = [];
    const handler = webhookReceiver({
      store,
      logger: (m) => messages.push(m),
      verify: () => {
        throw new Error('verifier exploded on key msg_1');
      },
    })(async () => new Response('handled'));
    const res = await handler(delivery());
    expect(res.status).toBe(500);
    const body = (await res.json()) as { code: string; detail: string };
    expect(body.code).toBe('configuration-error');
    expect(body.detail).toBe('the verify callback failed');
    expect(begins).toEqual([]);
    // Ruling 12: the malfunction is not silent, and the line carries neither the thrown error nor the id.
    expect(messages).toHaveLength(1);
    expect(messages[0]).not.toContain('msg_1');
    expect(messages[0]).not.toContain('exploded');
  });

  test('REQ-WH-2: the two configuration-error causes log independently and carry different details', async () => {
    const run = async (
      options: Parameters<typeof webhookReceiver>[0],
    ): Promise<{ detail: string; messages: string[] }> => {
      const messages: string[] = [];
      const handler = webhookReceiver({ ...options, logger: (m) => messages.push(m) })(
        async () => new Response('handled'),
      );
      let detail = '';
      for (let i = 0; i < 3; i += 1) {
        const res = await handler(delivery());
        expect(res.status).toBe(500);
        detail = ((await res.json()) as { detail: string }).detail;
      }
      return { detail, messages };
    };

    const unconfigured = await run({ store: countingStore().store });
    const failed = await run({
      store: countingStore().store,
      verify: () => {
        throw new Error('down');
      },
    });
    expect(unconfigured.detail).not.toBe(failed.detail);
    // Each cause has its own latch, so neither suppresses the other, and each still logs exactly once.
    expect(unconfigured.messages).toHaveLength(1);
    expect(failed.messages).toHaveLength(1);
    expect(unconfigured.messages[0]).not.toBe(failed.messages[0]);
  });

  test('REQ-WH-2: verify sees the raw body bytes and the request', async () => {
    const { store } = countingStore();
    const seen: string[] = [];
    const handler = webhookReceiver({
      store,
      verify: (req, body) => {
        seen.push(`${new URL(req.url).pathname}:${new TextDecoder().decode(body)}`);
        return false;
      },
    })(async () => new Response('handled'));
    await handler(delivery('{"b":2}'));
    expect(seen).toEqual(['/hooks/stripe:{"b":2}']);
  });

  test('REQ-WH-2: a verifiedMarker that was never set answers 401 and never calls begin', async () => {
    const { store, begins } = countingStore();
    const handler = webhookReceiver({ store, verifiedMarker: 'gateway' })(
      async () => new Response('handled'),
    );
    const res = await handler(delivery());
    expect(res.status).toBe(401);
    expect(begins).toEqual([]);
  });

  test('REQ-WH-2: a marker set by an upstream verifier passes the gate', async () => {
    const { store, begins } = countingStore();
    const handler = webhookReceiver({ store, verifiedMarker: 'gateway' })(
      async () => new Response('handled', { status: 202 }),
    );
    const req = delivery();
    markVerified(req, 'gateway');
    const res = await handler(req);
    expect(res.status).toBe(202);
    expect(begins).toHaveLength(1);
  });

  test('REQ-WH-2: a GET to a receiver with no verification still gets 500 configuration-error', async () => {
    const { store, begins } = countingStore();
    // The logger is injected purely so the one configuration line does not reach stderr and muddy the run.
    const handler = webhookReceiver({ store, logger: () => {} })(
      async () => new Response('handled'),
    );
    const res = await handler(new Request('https://example.test/hooks/stripe', { method: 'GET' }));
    expect(res.status).toBe(500);
    expect(((await res.json()) as { code: string }).code).toBe('configuration-error');
    expect(begins).toEqual([]);
  });

  test('REQ-WH-2: a custom onError still carries Cache-Control no-store on every gate problem', async () => {
    // Ruling 21: the receiver renders the gate's problems itself, so without the protocol header merge a
    // custom renderer produced cacheable 401, 500 and 413 responses while the bridge's own problems on the
    // same receiver were no-store. The renderer sets neither header, which is the whole point.
    const onError = (p: Problem): Response =>
      new Response(`custom:${p.code}`, { status: p.status });
    const unconfigured = webhookReceiver({
      store: countingStore().store,
      logger: () => {},
      onError,
    })(async () => new Response('handled'));
    const invalid = webhookReceiver({ store: countingStore().store, verify: () => false, onError })(
      async () => new Response('handled'),
    );
    const tooLarge = webhookReceiver({
      store: countingStore().store,
      verify: () => true,
      maxRequestBytes: 8,
      onError,
    })(async () => new Response('handled'));

    for (const [status, res] of [
      [500, await unconfigured(delivery())],
      [401, await invalid(delivery())],
      [413, await tooLarge(delivery('x'.repeat(64)))],
    ] as const) {
      expect(res.status).toBe(status);
      expect(await res.text()).toStartWith('custom:');
      expect(res.headers.get('Cache-Control')).toBe('no-store');
      // Ruling 20: the 401's challenge survives a custom renderer the same way.
      expect(res.headers.get('WWW-Authenticate')).toBe(status === 401 ? 'Signature' : null);
    }
  });

  test('REQ-WH-2: a 401 carries the RFC 9110 WWW-Authenticate challenge and no other problem does', async () => {
    // Ruling 20: RFC 9110 section 15.5.2 makes at least one challenge a MUST on a 401, and signature-invalid
    // is the only problem this door answers with one.
    const invalid = webhookReceiver({ store: countingStore().store, verify: () => false })(
      async () => new Response('handled'),
    );
    const unauthorized = await invalid(delivery());
    expect(unauthorized.status).toBe(401);
    expect(unauthorized.headers.get('WWW-Authenticate')).toBe('Signature');

    const unconfigured = webhookReceiver({ store: countingStore().store, logger: () => {} })(
      async () => new Response('handled'),
    );
    const tooLarge = webhookReceiver({
      store: countingStore().store,
      verify: () => true,
      maxRequestBytes: 8,
    })(async () => new Response('handled'));
    const missingId = webhookReceiver({ store: countingStore().store, verify: () => true })(
      async () => new Response('handled'),
    );
    for (const [status, res] of [
      [500, await unconfigured(delivery())],
      [413, await tooLarge(delivery('x'.repeat(64)))],
      [400, await missingId(delivery('{"a":1}', {}))],
    ] as const) {
      expect(res.status).toBe(status);
      expect(res.headers.get('WWW-Authenticate')).toBeNull();
    }
  });

  test('REQ-WH-2: an oversized body is 413 before verification and never calls begin', async () => {
    const { store, begins } = countingStore();
    const handler = webhookReceiver({ store, verify: () => false, maxRequestBytes: 8 })(
      async () => new Response('handled'),
    );
    const res = await handler(delivery('x'.repeat(64)));
    expect(res.status).toBe(413);
    expect(((await res.json()) as { code: string }).code).toBe('payload-too-large');
    expect(begins).toEqual([]);
  });
});

File: go/webhookmw/gate_test.go
package webhookmw_test

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"

	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/store/memory"
	"github.com/sns45/anyonce/go/webhookmw"
)

type countingStore struct {
	anyonce.Store
	mu     sync.Mutex
	begins []anyonce.Operation
}

func newCountingStore() *countingStore { return &countingStore{Store: memory.New()} }

func (c *countingStore) Begin(ctx context.Context, op anyonce.Operation, opts anyonce.BeginOptions) (anyonce.BeginOutcome, error) {
	c.mu.Lock()
	c.begins = append(c.begins, op)
	c.mu.Unlock()
	return c.Store.Begin(ctx, op, opts)
}

func (c *countingStore) count() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.begins)
}

func delivery(id, body string) *http.Request {
	r := httptest.NewRequest(http.MethodPost, "/hooks/stripe", strings.NewReader(body))
	if id != "" {
		r.Header.Set("webhook-id", id)
	}
	r.Header.Set("Content-Type", "application/json")
	return r
}

func handled() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("handled"))
	})
}

func TestGate(t *testing.T) {
	t.Run("REQ-WH-2: a receiver with neither Verify nor VerifiedMarker answers 500 configuration-error and never calls Begin", func(t *testing.T) {
		store := newCountingStore()
		var messages []string
		mw := webhookmw.New(store, webhookmw.Options{
			Logf: func(format string, args ...any) { messages = append(messages, fmt.Sprintf(format, args...)) },
		})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusInternalServerError {
			t.Fatalf("status = %d, want 500", rec.Code)
		}
		if got := rec.Header().Get("Content-Type"); got != "application/problem+json" {
			t.Fatalf("content type = %q", got)
		}
		var p webhookmw.Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Code != webhookmw.CodeConfigurationError || p.Status != 500 {
			t.Fatalf("problem = %+v", p)
		}
		if p.Title != "The webhook endpoint could not establish that this delivery is genuine" {
			t.Fatalf("title = %q", p.Title)
		}
		if p.Detail != "no verify callback or verifiedMarker is configured" {
			t.Fatalf("detail = %q", p.Detail)
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
		if len(messages) != 1 {
			t.Fatalf("logged %d times, want 1", len(messages))
		}
		if strings.Contains(messages[0], "msg_1") {
			t.Fatalf("the log line carries request data: %q", messages[0])
		}
	})

	t.Run("REQ-WH-2: the configuration error logs once however many requests arrive", func(t *testing.T) {
		store := newCountingStore()
		var messages []string
		mw := webhookmw.New(store, webhookmw.Options{
			Logf: func(format string, args ...any) { messages = append(messages, fmt.Sprintf(format, args...)) },
		})
		h := mw.Handler(handled())
		for i := range 3 {
			rec := httptest.NewRecorder()
			h.ServeHTTP(rec, delivery(fmt.Sprintf("msg_%d", i), `{"a":1}`))
			if rec.Code != http.StatusInternalServerError {
				t.Fatalf("request %d: status = %d, want 500", i, rec.Code)
			}
		}
		if len(messages) != 1 {
			t.Fatalf("logged %d times, want 1", len(messages))
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})

	t.Run("REQ-WH-2: the configuration check runs before the method filter, so a GET is 500 too", func(t *testing.T) {
		store := newCountingStore()
		var messages []string
		mw := webhookmw.New(store, webhookmw.Options{
			Logf: func(format string, args ...any) { messages = append(messages, fmt.Sprintf(format, args...)) },
		})
		rec := httptest.NewRecorder()
		r := httptest.NewRequest(http.MethodGet, "/hooks/stripe", nil)
		mw.Handler(handled()).ServeHTTP(rec, r)
		if rec.Code != http.StatusInternalServerError {
			t.Fatalf("status = %d, want 500", rec.Code)
		}
		if rec.Body.String() == "handled" {
			t.Fatal("the handler ran for a receiver that can never verify anything")
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})

	t.Run("REQ-WH-2: a Verify that returns false answers 401 signature-invalid and never calls Begin", func(t *testing.T) {
		store := newCountingStore()
		mw := webhookmw.New(store, webhookmw.Options{
			Verify: func(*http.Request, []byte) (bool, error) { return false, nil },
		})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusUnauthorized {
			t.Fatalf("status = %d, want 401", rec.Code)
		}
		var p webhookmw.Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Code != webhookmw.CodeSignatureInvalid {
			t.Fatalf("code = %q", p.Code)
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})

	t.Run("REQ-WH-2: Verify sees the raw body bytes and the request", func(t *testing.T) {
		store := newCountingStore()
		const payload = `{"nested":{"b":[1,2,3]},"unicode":"café"}`
		var seenBody []byte
		var seenMethod, seenPath, seenID string
		mw := webhookmw.New(store, webhookmw.Options{
			Verify: func(r *http.Request, body []byte) (bool, error) {
				seenBody = body
				seenMethod = r.Method
				seenPath = r.URL.Path
				seenID = r.Header.Get("webhook-id")
				return true, nil
			},
		})
		echo := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			read, err := io.ReadAll(r.Body)
			if err != nil {
				t.Errorf("the handler could not read the body: %v", err)
			}
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write(read)
		})
		rec := httptest.NewRecorder()
		mw.Handler(echo).ServeHTTP(rec, delivery("msg_1", payload))
		if string(seenBody) != payload {
			t.Fatalf("Verify saw body %q, want %q", seenBody, payload)
		}
		if seenMethod != http.MethodPost || seenPath != "/hooks/stripe" || seenID != "msg_1" {
			t.Fatalf("Verify saw %s %s with id %q", seenMethod, seenPath, seenID)
		}
		// The gate reads the body to hand it to Verify, so it has to put the bytes back for the handler.
		if rec.Body.String() != payload {
			t.Fatalf("the handler read %q, want %q", rec.Body.String(), payload)
		}
	})

	t.Run("REQ-WH-2: a Verify that returns an error is 500 configuration-error and never calls Begin", func(t *testing.T) {
		store := newCountingStore()
		var messages []string
		mw := webhookmw.New(store, webhookmw.Options{
			Logf: func(format string, args ...any) { messages = append(messages, fmt.Sprintf(format, args...)) },
			Verify: func(*http.Request, []byte) (bool, error) {
				return false, fmt.Errorf("the key service is down, key msg_1")
			},
		})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusInternalServerError {
			t.Fatalf("status = %d, want 500", rec.Code)
		}
		var p webhookmw.Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Code != webhookmw.CodeConfigurationError || p.Status != 500 {
			t.Fatalf("problem = %+v", p)
		}
		if p.Detail != "the verify callback failed" {
			t.Fatalf("detail = %q", p.Detail)
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
		// Ruling 12: the malfunction is not silent, and ruling 12 again: the line carries neither the
		// verifier's own error text nor anything from the request.
		if len(messages) != 1 {
			t.Fatalf("logged %d times, want 1", len(messages))
		}
		for _, leak := range []string{"msg_1", "key service"} {
			if strings.Contains(messages[0], leak) {
				t.Fatalf("the log line carries %q: %q", leak, messages[0])
			}
		}
	})

	t.Run("REQ-WH-2: the two configuration-error causes log independently and carry different details", func(t *testing.T) {
		detailFor := func(t *testing.T, opts webhookmw.Options) (string, []string) {
			t.Helper()
			var messages []string
			opts.Logf = func(format string, args ...any) { messages = append(messages, fmt.Sprintf(format, args...)) }
			mw := webhookmw.New(newCountingStore(), opts)
			h := mw.Handler(handled())
			var p webhookmw.Problem
			for range 3 {
				rec := httptest.NewRecorder()
				h.ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
				if rec.Code != http.StatusInternalServerError {
					t.Fatalf("status = %d, want 500", rec.Code)
				}
				if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
					t.Fatal(err)
				}
			}
			return p.Detail, messages
		}

		unconfiguredDetail, unconfiguredMessages := detailFor(t, webhookmw.Options{})
		failedDetail, failedMessages := detailFor(t, webhookmw.Options{
			Verify: func(*http.Request, []byte) (bool, error) { return false, fmt.Errorf("down") },
		})
		if unconfiguredDetail == failedDetail {
			t.Fatalf("both causes report the same detail %q", unconfiguredDetail)
		}
		// Each cause has its own latch, so neither can suppress the other, and each still logs exactly once.
		if len(unconfiguredMessages) != 1 || len(failedMessages) != 1 {
			t.Fatalf("logged %d and %d times, want 1 each", len(unconfiguredMessages), len(failedMessages))
		}
		if unconfiguredMessages[0] == failedMessages[0] {
			t.Fatalf("both causes log the same line %q", unconfiguredMessages[0])
		}
	})

	t.Run("REQ-WH-2: a configured receiver passes a non-POST straight to next without verifying", func(t *testing.T) {
		store := newCountingStore()
		verifyRan := false
		mw := webhookmw.New(store, webhookmw.Options{
			Verify: func(*http.Request, []byte) (bool, error) {
				verifyRan = true
				return false, nil
			},
		})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, httptest.NewRequest(http.MethodGet, "/hooks/stripe", nil))
		// This is the one path that reaches next without verification. It is intended, it mirrors the HTTP
		// door's method filter, and it is safe only because a method the receiver does not claim is never
		// deduplicated either: nothing about it reaches the store.
		if rec.Code != http.StatusOK || rec.Body.String() != "handled" {
			t.Fatalf("status = %d, body = %q, want 200 and handled", rec.Code, rec.Body.String())
		}
		if verifyRan {
			t.Fatal("Verify ran for a method the receiver does not apply to")
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})

	t.Run("REQ-WH-2: a VerifiedMarker that was never set answers 401 and never calls Begin", func(t *testing.T) {
		store := newCountingStore()
		mw := webhookmw.New(store, webhookmw.Options{VerifiedMarker: "stripe"})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusUnauthorized {
			t.Fatalf("status = %d, want 401", rec.Code)
		}
		var p webhookmw.Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Code != webhookmw.CodeSignatureInvalid {
			t.Fatalf("code = %q", p.Code)
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})

	t.Run("REQ-WH-2: a marker set by an upstream verifier passes the gate", func(t *testing.T) {
		matched := newCountingStore()
		mw := webhookmw.New(matched, webhookmw.Options{VerifiedMarker: "stripe"})
		upstream := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx := webhookmw.MarkVerified(r.Context(), "stripe")
			mw.Handler(handled()).ServeHTTP(w, r.WithContext(ctx))
		})
		rec := httptest.NewRecorder()
		upstream.ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusOK {
			t.Fatalf("status = %d, want 200", rec.Code)
		}
		if rec.Body.String() != "handled" {
			t.Fatalf("body = %q, want handled", rec.Body.String())
		}
		// One, now that key resolution has landed: a delivery that passes the gate is deduplicated, so the
		// verified request reaches the store exactly once.
		if matched.count() != 1 {
			t.Fatalf("Begin was called %d times", matched.count())
		}

		// A different marker name is not the one that was set, and it gets its own store so the two halves
		// cannot borrow each other's count.
		mismatched := newCountingStore()
		other := webhookmw.New(mismatched, webhookmw.Options{VerifiedMarker: "github"})
		rec = httptest.NewRecorder()
		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx := webhookmw.MarkVerified(r.Context(), "stripe")
			other.Handler(handled()).ServeHTTP(w, r.WithContext(ctx))
		}).ServeHTTP(rec, delivery("msg_1", `{"a":1}`))
		if rec.Code != http.StatusUnauthorized {
			t.Fatalf("status for the wrong marker = %d, want 401", rec.Code)
		}
		if mismatched.count() != 0 {
			t.Fatalf("Begin was called %d times", mismatched.count())
		}
	})

	t.Run("REQ-WH-2: a 401 carries the RFC 9110 WWW-Authenticate challenge and no other problem does", func(t *testing.T) {
		// Ruling 20: RFC 9110 section 15.5.2 makes at least one challenge a MUST on a 401, and
		// signature-invalid is the only problem this door answers with one.
		unauthorized := httptest.NewRecorder()
		webhookmw.New(newCountingStore(), webhookmw.Options{
			Verify: func(*http.Request, []byte) (bool, error) { return false, nil },
		}).Handler(handled()).ServeHTTP(unauthorized, delivery("msg_1", `{"a":1}`))
		if unauthorized.Code != http.StatusUnauthorized {
			t.Fatalf("status = %d, want 401", unauthorized.Code)
		}
		if got := unauthorized.Header().Get("WWW-Authenticate"); got != "Signature" {
			t.Fatalf("WWW-Authenticate = %q, want Signature", got)
		}

		// A 500 configuration-error, a 413 payload-too-large and a 400 missing-key are not 401s and carry
		// no challenge.
		unconfigured := httptest.NewRecorder()
		webhookmw.New(newCountingStore(), webhookmw.Options{Logf: func(string, ...any) {}}).
			Handler(handled()).ServeHTTP(unconfigured, delivery("msg_1", `{"a":1}`))
		tooLarge := httptest.NewRecorder()
		webhookmw.New(newCountingStore(), webhookmw.Options{MaxRequestBytes: 8, Verify: func(*http.Request, []byte) (bool, error) { return true, nil }}).
			Handler(handled()).ServeHTTP(tooLarge, delivery("msg_1", strings.Repeat("x", 64)))
		missingID := httptest.NewRecorder()
		webhookmw.New(newCountingStore(), webhookmw.Options{Verify: func(*http.Request, []byte) (bool, error) { return true, nil }}).
			Handler(handled()).ServeHTTP(missingID, delivery("", `{"a":1}`))
		for _, tc := range []struct {
			name string
			want int
			rec  *httptest.ResponseRecorder
		}{
			{"configuration-error", http.StatusInternalServerError, unconfigured},
			{"payload-too-large", http.StatusRequestEntityTooLarge, tooLarge},
			{"missing-key", http.StatusBadRequest, missingID},
		} {
			if tc.rec.Code != tc.want {
				t.Fatalf("%s: status = %d, want %d", tc.name, tc.rec.Code, tc.want)
			}
			if got := tc.rec.Header().Get("WWW-Authenticate"); got != "" {
				t.Fatalf("%s: WWW-Authenticate = %q, want none", tc.name, got)
			}
		}
	})

	t.Run("REQ-WH-2: an oversized body is 413 before verification and never calls Begin", func(t *testing.T) {
		store := newCountingStore()
		verified := false
		mw := webhookmw.New(store, webhookmw.Options{
			MaxRequestBytes: 8,
			Verify: func(*http.Request, []byte) (bool, error) {
				verified = true
				return true, nil
			},
		})
		rec := httptest.NewRecorder()
		mw.Handler(handled()).ServeHTTP(rec, delivery("msg_1", strings.Repeat("x", 64)))
		if rec.Code != http.StatusRequestEntityTooLarge {
			t.Fatalf("status = %d, want 413", rec.Code)
		}
		var p webhookmw.Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Code != webhookmw.CodePayloadTooLarge {
			t.Fatalf("code = %q", p.Code)
		}
		if verified {
			t.Fatal("Verify ran for a body the gate had already rejected")
		}
		if store.count() != 0 {
			t.Fatalf("Begin was called %d times", store.count())
		}
	})
}

func TestMarkers(t *testing.T) {
	t.Run("REQ-WH-2: IsVerified reports false for a context no verifier touched", func(t *testing.T) {
		if webhookmw.IsVerified(context.Background(), "stripe") {
			t.Fatal("an untouched context claims to be verified")
		}
	})

	t.Run("REQ-WH-2: MarkVerified keeps the markers already on the context", func(t *testing.T) {
		ctx := webhookmw.MarkVerified(context.Background(), "stripe")
		derived := webhookmw.MarkVerified(ctx, "github")
		for _, marker := range []string{"stripe", "github"} {
			if !webhookmw.IsVerified(derived, marker) {
				t.Fatalf("the derived context lost the %q marker", marker)
			}
		}
		// Marking a derived context never reaches back into the parent.
		if webhookmw.IsVerified(ctx, "github") {
			t.Fatal("the parent context gained a marker set on a child")
		}
	})
}
